File: solutions/elixir/word-count/4/word-count.exs
defmodule Words do
  def count(phrase), do: normalize_words(phrase) |> count_words

  # returns a list of the lowercased words (without punctuation)
  defp normalize_words(phrase) do
    Regex.scan(%r/\w+/, String.downcase(phrase)) |> List.concat
  end

  # count the occurrences of words in the phrase
  defp count_words(words), do: Enum.reduce(words, HashDict.new, count(&1, &2))

  # helper function to increment a counter
  defp increment(current_value), do: current_value + 1

  # helper function to update the HashDict of counts
  defp count(word, acc), do: HashDict.update(acc, word, 1, increment(&1))
end

File: solutions/javascript/word-count/1/word-count.js
"use strict";
var Words = (function() {
  function Words(wordString) {
    var _wordString = wordString.toLowerCase();
    var counts = _wordString.match(/\b[a-z0-9]+\b/gi).
      reduce(function(collection, word) {
        collection[word] = (collection[word] || 0) + 1;
        return collection;
      }, {});
    this.count = counts;
  }
  return Words;
})();

module.exports = Words;